EventGenerator/src/PionFilter_module.cc

```diff
@@ -44,60 +51,85 @@ namespace mu2e {
       int diagLevel_;
       double tmin_;
       double tmax_;
+      int maxPions_;
       int processCode_;
+      art::ProductToken<SimParticleCollection> const simsToken_;
       bool isNull_;
-      float _totalweight = 0;
-      float _selectedweight = 0;
-      int _ntot = 0;
-      int _nsel = 0;
+      SumOfWeights total_;
+      SumOfWeights selected_;
   };
 
   PionFilter::PionFilter(const art::EDFilter::Table<Config>& config) :
      EDFilter{config}
     , diagLevel_{config().diagLevel()}
-    , tmin_{config().tmin()}
-    , tmax_{config().tmax()}
     , processCode_{config().processCode()}
+    , simsToken_{consumes<SimParticleCollection>(config().simCollTag())}
     , isNull_{config().isNull()}
   {
+    if(!config().tmin(tmin_)) tmin_ = -1.e10;
+    if(!config().tmax(tmax_)) tmax_ =  1.e10;
+    if(!config().maxPions(maxPions_)) maxPions_ = -1;
+    produces<SumOfWeights, art::InSubRun>("total");
+    produces<SumOfWeights, art::InSubRun>("selected");
   }
 
   void PionFilter::beginJob(){
       art::ServiceHandle<art::TFileService> tfs;
-
-}
+  }
 
   bool PionFilter::filter(art::Event& evt) {
       if(isNull_) return true;
       bool passed = false;
-      std::vector<art::Handle<SimParticleCollection>> vah = evt.getMany<SimParticleCollection>();
-      for (auto const& ah : vah) { //always one collection
-        for(const auto& aParticle : *ah){
-          art::Ptr<SimParticle> pp(ah, aParticle.first.asUint());
+      const auto simh = evt.getValidHandle<SimParticleCollection>(simsToken_);
+      const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+      const std::vector<int> decayOffCodes = {PDGCode::pi_plus, PDGCode::pi_minus};
```

Comment:
If we made decayOffCodes a subrun data product of the generation this would allow testing  how the input was generated. Again outside scope of this PR.